### PR TITLE
[#2697] improvement(core): Support different entity with the same name under a namespace

### PR DESCRIPTION
**Pull request by yuqi1129**
### What changes were proposed in this pull request?

Support store entities with the same name under a namespace. 

### Why are the changes needed?

It's a necessary feature for Gravitino. 

Fix: #2697 

### Does this PR introduce _any_ user-facing change?

N/A.

### How was this patch tested?

UTs

### CODE REVIEW

File: core/src/main/java/com/datastrato/gravitino/Entity.java [hunk 1]
```diff
@@ -41,6 +43,35 @@ public static EntityType fromShortName(String shortName) {
...
+        case TABLE:
+          return ImmutableList.of(METALAKE, CATALOG, SCHEMA);
+        case FILESET:
+          return ImmutableList.of(METALAKE, CATALOG, SCHEMA);
+        case TOPIC:
+          return ImmutableList.of(METALAKE, CATALOG, SCHEMA);
+        case COLUMN:
+          return ImmutableList.of(METALAKE, CATALOG, SCHEMA, TABLE);
+        default:
+          throw new IllegalArgumentException("Unknown entity type: " + entityType);
```
Comment (jerryshao) [inline core/src/main/java/com/datastrato/gravitino/Entity.java:72]: Can you please rebase your PR based on @qqqttt123 newly added USER and GROUP support.

**Comment by jerryshao** [comment]
I think the key thing here is to make this change backward compatible, otherwise there's no way for users to upgrade.

**Comment by yuqi1129** [comment]
> I think the key thing here is to make this change backward compatible, otherwise there's no way for users to upgrade.

Yeah, that is why I'm hesitant about it.

File: core/src/main/java/com/datastrato/gravitino/storage/kv/KvEntityStore.java [hunk 2]
```diff
@@ -199,72 +189,15 @@ public <E extends Entity & HasIdentifier> E update(
 
           // Update the name mapping
           nameMappingService.updateName(
-              generateKeyForMapping(ident), generateKeyForMapping(updatedE.nameIdentifier()));
+              generateKeyForMapping(ident, entityType, nameMappingService),
+              generateKeyForMapping(updatedE.nameIdentifier(), entityType, nameMappingService));
```
Comment (jerryshao) [inline core/src/main/java/com/datastrato/gravitino/storage/kv/KvEntityStore.java:193]: Only changing here is enough, do you need to change put/get also?

Comment (yuqi1129) [reply core/src/main/java/com/datastrato/gravitino/storage/kv/KvEntityStore.java:193]: all have been changed through method `concatIdAndName` and `put/get/update` will use this method to generate key format.

File: core/src/main/java/com/datastrato/gravitino/Entity.java [hunk 3]
```diff
@@ -43,6 +45,37 @@ public static EntityType fromShortName(String shortName) {
...
+     * @param entityType The entity type for which to get the parent entity types.
+     * @return The parent entity types of the given entity type.
+     */
+    public static List<EntityType> getParentEntityTypes(EntityType entityType) {
+      switch (entityType) {
+        case METALAKE:
+          return ImmutableList.of();
+        case CATALOG:
+        case USER:
+        case GROUP:
```
Comment (yuqi1129) [inline core/src/main/java/com/datastrato/gravitino/Entity.java:63]: @qqqttt123 , Can you confirm if these changes are acceptable?

Comment (qqqttt123) [reply core/src/main/java/com/datastrato/gravitino/Entity.java:63]: Now, My nameIdentifier is `metalake."system"."user".user1` and `metalake."system"."group".group1`.

**Comment by yuqi1129** [comment]
> > I think the key thing here is to make this change backward compatible, otherwise there's no way for users to upgrade.
> 
> Yeah, that is why I'm hesitant about it.

I have updated the code and added support for backward compatibility. Please take a look if you have time. Thanks.

File: core/src/main/java/com/datastrato/gravitino/storage/kv/BinaryEntityEncoderUtil.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        prefixes.add(replacePrefixTypeInfo(encode, SCHEMA.getShortName()));
+        prefixes.add(replacePrefixTypeInfo(encode, TABLE.getShortName()));
+        prefixes.add(replacePrefixTypeInfo(encode, FILESET.getShortName()));
+        break;
+      case SCHEMA:
+        prefixes.add(replacePrefixTypeInfo(encode, TABLE.getShortName()));
+        prefixes.add(replacePrefixTypeInfo(encode, FILESET.getShortName()));
+        break;
+      case TABLE:
+      case FILESET:
```
Comment (jerryshao) [inline core/src/main/java/com/datastrato/gravitino/storage/kv/BinaryEntityEncoderUtil.java:160]: Should you also add USER and GROUP here?

Comment (qqqttt123) [reply core/src/main/java/com/datastrato/gravitino/storage/kv/BinaryEntityEncoderUtil.java:160]: If we add USER and GROUP here, USER and GROUP will block the non-cascade deletion of metalake.

File: core/src/main/java/com/datastrato/gravitino/storage/kv/BinaryEntityEncoderUtil.java [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+   * @return concatenated string that used in the id-name mapping.
+   */
+  public static String concatIdAndName(long[] namespaceIds, String name, EntityType type) {
+    String context =
+        Joiner.on(NAMESPACE_SEPARATOR)
+            .join(
+                Arrays.stream(namespaceIds).mapToObj(String::valueOf).collect(Collectors.toList()));
+    // We need to make it backward compatible, so we need to check if the name is already prefixed.
+    String mappingName =
+        VERSION_0_4_0_COMPATIBLE_ENTITY_TYPES.contains(type)
```
Comment (jerryshao) [inline core/src/main/java/com/datastrato/gravitino/storage/kv/BinaryEntityEncoderUtil.java:114]: Better to change to `VERSION_0_4_COMPATIBLE_ENTITY_TYPES` to ignore the minor version.

Comment (jerryshao) [reply core/src/main/java/com/datastrato/gravitino/storage/kv/BinaryEntityEncoderUtil.java:114]: @yuqi1129 please fix this issue, thanks!

**Review by jerryshao — APPROVED** [review]
LGTM. Thanks for your work @yuqi1129

**Pull request merged**